### ISSUE TITLE
perf: tape backward + optimizer is 15x forward on 25M-param Paraformer (Persistent-tape cached-chain fast-path under-delivers)

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#433]
## Summary

Profiled a paper-faithful SenseVoice-Small (Du et al. 2024 — 12 Conformer encoder layers + 6 non-AR decoder layers, 25.4 M params, 176 layers total) running a single training step at `[1, 64, 32]` input through `[1, 64, 25000]` output on the current main branch. The backward + optimizer phase consumes **94% of the training step wall time** at steady state. Net: 100 memorization iters (the default for `LossStrictlyDecreasesOnMemorizationTask` invariant) project to ~460 s vs the 180 s xUnit timeout budget.

## Profile numbers

CPU-only (`AiDotNetEngine.ResetToCpu()`), `float`, Adam optimizer, MSE loss, fresh master + AiDotNet#1421's `BN→LN` fix:

```
Ctor + InitializeLayers:        49 ms
Predict #1 (lazy init):       2269 ms
Predict #2 (warm):             304 ms
Predict avg (5 iters):         292 ms
ForwardForTraining avg (5):    273 ms (= forward inside Train)

Train #1 (warm-up):           5616 ms
Train #2:                     4596 ms
Train #3:                     4909 ms
Train #4:                     4373 ms

=> Backward + optimizer ≈ Train − ForwardForTraining ≈ 4500 − 273 ≈ 4227 ms (94% of step)
```

For reference, PyTorch's typical backward/forward ratio on CPU is 1.5×–2×. The observed 15×–18× ratio here suggests a structural overhead in the tape backward dispatch.

## Suspected gaps

Per the `NeuralNetworkBase.TrainWithTape` comments at line 5553:
> Persistent = true gates the AutoTrainingCompiler fast path. With Persistent = true, after the first training step the compiler records the forward op pattern; on subsequent steps with a matching pattern, ComputeGradients replays a compiled CompiledBackwardGraph instead of walking the tape entry list + dispatching dictionary-keyed gradient lookups per op. Profiling (dotnet-trace + GC.GetTotalAllocatedBytes) showed gradient-tape backward dominating training-step time on paper-scale CNNs (~838 ms / call out of ~1.3 s VGG11 Train, ~73 % of step wall-time; similar fraction for ResNet50). Per-step allocations also drop because the compiled backward keeps tensors in a flat indexed array rather than the dictionary-of-tensor-refs the tape-walk path uses.

Expected: on Train #2 onward, `_cachedDelegateChain` should replay and deliver ~73% backward speedup → step time should drop from ~5.6 s to ~1.5–2 s.

Observed: only ~20% speedup (5.6 → 4.6 s). The fast path either doesn't engage for this pattern, or engages but is itself heavy.

Hot candidates worth dotnet-trace runs against this profile harness:

1. **Cached-chain engagement check**: does `_cachedDelegateChain` actually become non-null after Train #1 for SenseVoice's layer pattern (176 layers, 25M params, mixed Dense+MHA+LayerNorm+Dropout)? If yes, why is replay 4× slower than the VGG11 baseline that produced the 73% speedup metric?

2. **Per-op backward kernel dispatch**: even on cached-chain replay each `step.OutputBuffer` requires a backward kernel dispatch. For SenseVoice there are ~600+ ops on the tape (176 layers × ~3-4 ops avg). At 4500 ms / 600 ops = ~7.5 ms/op average — way above the kernel's actual FMA cost on a `[1, 64, 512]` tile.

3. **MHA backward dispatch**: 24 MHA layers each running a tape-walked backward. Each MHA backward needs to: dQ, dK, dV via 4 GEMMs each. Possibly the BLAS dispatch path isn't engaging for the 24x16-tile shape these GEMMs produce.

4. **Allocations**: the comment at line 5562 mentions allocations dropping under cached-chain replay. Is that actually happening for SenseVoice's tape pattern? A GC.GetTotalAllocatedBytes diff per step would confirm.

## Repro

`tests/AiDotNet.Tests/Performance/SenseVoiceTrainStepProfile.cs` in AiDotNet PR #1421 — runs the model + emits the table above via `ITestOutputHelper`. Self-contained, paper-faithful defaults, ~25 s wall time for the full run. Use it as the fixture for any perf-fix verification.

## Downstream

Blocks `SenseVoiceTests.LossStrictlyDecreasesOnMemorizationTask` and the same test for SenseVoiceLarge / SeACo / ParaformerLarge / Paraformer (all share `CreateDefaultParaformerLayers`). Test was failing at the shape-collapse boundary (AiDotNet#1421 fixed that) and now times out on the perf gap.
